builtin/v13/migration/market.go

```diff
@@ -0,0 +1,409 @@
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load states array: %w", err)
+	}
+
+	newStateArray, err := adt13.MakeEmptyArray(ctxStore, market13.StatesAmtBitwidth)
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load states array: %w", err)
+	}
+
+	proposalsArr, err := adt.AsArray(ctxStore, proposals, market12.ProposalsAmtBitwidth)
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load proposals array: %w", err)
+	}
+
+	var oldState market12.DealState
+	var newState market13.DealState
+
+	err = oldStateArray.ForEach(&oldState, func(i int64) error {
+		deal := abi.DealID(i)
+
+		newState.SlashEpoch = oldState.SlashEpoch
+		newState.LastUpdatedEpoch = oldState.LastUpdatedEpoch
+		newState.SectorStartEpoch = oldState.SectorStartEpoch
+		newState.SectorNumber = 0 // non- -1 slashed epoch
+
+		var proposal market12.DealProposal
+		ok, err := proposalsArr.Get(uint64(i), &proposal)
+		if err != nil {
+			return xerrors.Errorf("failed to get proposal: %w", err)
+		}
+
+		if !ok {
+			return xerrors.Errorf("failed to find proposal for deal ID %d", i)
+		}
+
+		// FIP: For each unexpired deal state object in the market actor state that has a terminated epoch set to -1:
+		if oldState.SlashEpoch == -1 && proposal.EndEpoch >= UpgradeHeight {
+			// FIP: find the corresponding deal proposal object and extract the provider's actor ID;
+			// - we do this by collecting all dealIDs in providerSectors in miner migration
+
+			// in the provider's miner state, find the ID of the sector with the corresponding deal ID in sector metadata;
+			sid, ok := m.providerSectors.dealToSector[deal]
+			if ok {
+				newState.SectorNumber = sid.Number // FIP: set the new deal state object's sector number to the sector ID found;
+			} else {
+				return xerrors.Errorf("sector not found for unexpired deal %d", i)
+			}
+		}
+
+		if err := newStateArray.Set(uint64(deal), &newState); err != nil {
+			return xerrors.Errorf("failed to append new state: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to iterate states: %w", err)
+	}
+
+	newStateArrayRoot, err := newStateArray.Root()
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to get newStateArrayRoot: %w", err)
+	}
+
+	outProviderSectors, err := adt.MakeEmptyMap(ctxStore, market13.ProviderSectorsHamtBitwidth)
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to create empty map: %w", err)
+	}
+
+	for miner, sectors := range m.providerSectors.minerToSectorToDeals {
+		actorSectors, err := adt.MakeEmptyMap(ctxStore, market13.ProviderSectorsHamtBitwidth)
+		if err != nil {
+			return cid.Undef, cid.Undef, xerrors.Errorf("failed to create empty map: %w", err)
+		}
+
+		var sectorDealIDs market13.SectorDealIDs
+		for sector, deals := range sectors {
+			sectorDealIDs = deals
```

Comment:
I can't wait for the fixed loop to ship by default. This is one of the dumbest go design decisions.